rust/src/message/payload/sum.rs

```diff
@@ -0,0 +1,293 @@
+    /// `Ok(())` this method is guaranteed not to panic.
+    pub fn ephm_pk(&self) -> &'a [u8] {
+        &self.inner.as_ref()[EPHM_PK_RANGE]
+    }
+
+    /// Get a reference to the sum signature field
+    ///
+    /// # Panic
+    ///
+    /// This may panic if the underlying buffer does not represent a
+    /// valid update. If `self.check_buffer_length()` returned
+    /// `Ok(())` this method is guaranteed not to panic.
+    pub fn sum_signature(&self) -> &'a [u8] {
+        &self.inner.as_ref()[SUM_SIGNATURE_RANGE]
+    }
+}
+
+/// High level representation of a sum message. These messages are
+/// sent by sum partipants during the sum phase.
+///
+/// # Examples
+///
+/// ## Decoding a message
+///
+/// ```rust
+/// # use xain_fl::{crypto::ByteObject, message::{FromBytes, SumOwned}, ParticipantTaskSignature, SumParticipantEphemeralPublicKey};
+/// let signature = vec![0x11; 64];
+/// let ephm_pk = vec![0x22; 32];
+/// let bytes = [signature.as_slice(), ephm_pk.as_slice()].concat();
+/// let parsed = SumOwned::from_bytes(&bytes).unwrap();
+/// let expected = SumOwned {
+///     sum_signature: ParticipantTaskSignature::from_slice(&signature[..]).unwrap(),
+///     ephm_pk: SumParticipantEphemeralPublicKey::from_slice(&ephm_pk[..]).unwrap(),
+/// };
+/// assert_eq!(parsed, expected);
+/// ```
+///
+/// ## Encoding a message
+///
+/// ```rust
+/// # use xain_fl::{crypto::ByteObject, message::{ToBytes, Sum}, ParticipantTaskSignature, SumParticipantEphemeralPublicKey};
+/// let sum_signature = ParticipantTaskSignature::from_slice(vec![0x11; 64].as_slice()).unwrap();
+/// let ephm_pk = SumParticipantEphemeralPublicKey::from_slice(vec![0x22; 32].as_slice()).unwrap();
+/// let msg = Sum {
+///     sum_signature,
+///     ephm_pk,
+/// };
+/// // we need a 96 bytes long buffer to serialize that message
+/// assert_eq!(msg.buffer_length(), 96);
+/// // create a buffer with enough space and encode the message
+/// let mut buf = vec![0xff; 96];
+/// msg.to_bytes(&mut buf);
+///
+/// assert_eq!(buf, [vec![0x11; 64].as_slice(), vec![0x22; 32].as_slice()].concat());
+/// ```
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Sum {
+    /// Signature of the round seed and the word "sum", used to
+    /// determine whether a participant is selected for the sum task
+    pub sum_signature: ParticipantTaskSignature,
+    /// A public key generated by a sum participant for the current
+    /// round.
+    pub ephm_pk: SumParticipantEphemeralPublicKey,
+}
+
+pub type SumOwned = Sum;
+
+impl ToBytes for Sum {
+    fn buffer_length(&self) -> usize {
+        EPHM_PK_RANGE.end
+    }
+
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+        let mut writer = SumBuffer::new(buffer.as_mut()).unwrap();
+        writer
+            .sum_signature_mut()
+            .copy_from_slice(self.sum_signature.as_slice());
+
+        writer
+            .ephm_pk_mut()
+            .copy_from_slice(self.ephm_pk.as_slice());
+    }
```

Comment:
general API question: is there any particular reason `ParticipantTaskSignature` and `SumParticipantEphemeralPublicKey` don't `impl ToBytes`? then you could implement this more uniformly.

Reply:
They do implement `ToBytes` actually:

```rust
impl<T> ToBytes for T
where
    T: ByteObject,
{
    fn buffer_length(&self) -> usize {
        self.as_slice().len()
    }

    fn to_bytes<U: AsMut<[u8]>>(&self, buffer: &mut U) {
        buffer.as_mut().copy_from_slice(self.as_slice())
    }
}
```
and you're right this should rewritten to use `to_bytes` instead.